### PR TITLE
[clang][CodeGen] Shift relink option implementation away from module cloning

### PR DESCRIPTION
**Pull request by lamb-j**
We recently implemented a new option allowing relinking of bitcode modules via the "-mllvm -relink-builtin-bitcode-postop"
option.

This implementation relied on llvm::CloneModule() in order to pass copies to modules and preserve the original modules for later relinking. However, cloning modules has been found to be prohibitively expensive, significantly increasing compilation time for large bitcode libraries.

In this patch, we shift the relink option implementation to instead link the original modules initially, and reload modules from the file system if relinking is requested. This approach results in significantly reduced overhead.

We accomplish this by creating a new ReloadModules() routine that can be called from a BackendConsumer class, to mimic the behavior of ASTConsumer's loadLinkModules(), but without access to the CompilerInstance.

Because loading the bitcodes from the filesystem requires access to the FileManager class, we also forward a reference to the CompilerInstance class to the BackendConsumer. This mirrors what is already done for several CompilerInstance members, such as TargetOptions and CodeGenOptions.

Finally, we needed to add a const specifier to the FileManager::getBufferForFile() routine to allow it to be called using the const reference returned from CompilerInstance::getFileManager()

### CODE REVIEW

**Review by yxsamliu — APPROVED** [review]
LGTM. Thanks

**Review by jhuber6 — APPROVED** [review]
Looks fine.

File: clang/lib/CodeGen/CodeGenAction.cpp [hunk 1]
```diff
@@ -257,37 +281,19 @@ bool BackendConsumer::LinkInModules(llvm::Module *M, bool ShouldLinkFiles) {
     CurLinkModule = LM.Module.get();
     bool Err;
```
Comment (AaronHLiu) [inline clang/lib/CodeGen/CodeGenAction.cpp:282]: Got error:
```
clang/lib/CodeGen/CodeGenAction.cpp:282:10: error: variable 'Err' set but not used [-Werror,-Wunused-but-set-variable]
    bool Err;
         ^
1 error generated.
```

Comment (lamb-j) [reply clang/lib/CodeGen/CodeGenAction.cpp:282]: Oversight on my part. Should be fixed by https://github.com/llvm/llvm-project/pull/81777

Comment (AaronHLiu) [reply clang/lib/CodeGen/CodeGenAction.cpp:282]: Thanks!

**Pull request merged**